### PR TITLE
Return InferenceData objects when sampling and forward sampling

### PR DESCRIPTION
**Pull request by aloctavodia**
I still need to update documentation and tests

### CODE REVIEW

File: pymc4/inference/utils.py [hunk 1]
```diff
@@ -29,23 +29,42 @@ def initialize_sampling_state(
...
+    observed_data=None,
+    prior_predictive=None,
+    posterior_predictive=None,
+):
     """
     Tensorflow to Arviz trace convertor.
 
-    Convert a PyMC4 trace as returned by sample() to an ArviZ trace object
-    that can be passed to e.g. arviz.plot_trace().
+    Creates an ArviZ's InferenceData object with inference, prediction and/or sampling data
```
Comment (ColCarroll) [inline pymc4/inference/utils.py]: It would be nice if we made pymc4 support InferenceData independently: that is, directly supporting the schema defined there.

We'll should still use ArviZ as a way to compute diagnostics and visualizations, but the further down we push compatibility, the better, in my opinion!

**Comment by aloctavodia** [comment]
one more commit and we can merge! I still have to fix a couple of details :-)

File: tests/test_forward_sampling.py [hunk 2]
```diff
@@ -8,7 +8,7 @@
 from pymc4.flow.executor import EvaluationError
 
 
-@pytest.fixture(scope="module", params=[(), (1,), (2,), (1, 1), (7, 3)], ids=str)
+@pytest.fixture(scope="module", params=[(1, 10), (1, 1), (1, 2), (1, 1), (3, 7)], ids=str)
```
Comment (lucianopaz) [inline tests/test_forward_sampling.py]: Why did you remove the `()` sample shape? I think it's important to test it to see if we get the correct core shape or for some buggy reason, we are inventing new dimension

File: pymc4/inference/utils.py (hunk 1 above)
Comment (aloctavodia) [reply pymc4/inference/utils.py]: Yeah, I tough about this and the opposite alternative (implementing `az.from_pymc4`), and then decide on this PR, that is somehow an intermediate option. I do not have a clear opinion on what should be the best choice :-)

File: tests/test_forward_sampling.py [hunk 3]
```diff
@@ -189,19 +192,23 @@ def test_sample_posterior_predictive(posterior_predictive_fixture):
...
 
 def test_sample_ppc_var_names(model_fixture):
     model, observed = model_fixture
-    trace = {
-        "model/sd": tf.convert_to_tensor(np.array(1.0, dtype="float32")),
-        "model/y": tf.convert_to_tensor(observed),
-    }
+    trace = pm.inference.utils.trace_to_arviz(
+        {
+            "model/sd": tf.ones((10, 1), dtype="float32"),
```
Comment (lucianopaz) [inline tests/test_forward_sampling.py]: Why is the singleton axis needed here?

Comment (aloctavodia) [reply tests/test_forward_sampling.py]: because we have one chain

File: pymc4/forward_sampling.py [hunk 4]
```diff
@@ -112,24 +109,24 @@ def sample_prior_predictive(
...
+    downstream computations
 
     >>> prior_samples = sample_prior_predictive(
     ...     model2(), sample_shape=(20,), sample_from_observed=True
     ... )
     >>> np.allclose(np.mean(prior_samples["model2/y"]), 1)
     False
     >>> prior_samples["model2/y"].shape
-    (20,)
+    (1, 20)
```
[the remark this answers is not in the document]
Comment (aloctavodia) [reply pymc4/forward_sampling.py]: I think this is because the first dimension is the number of chains.

File: pymc4/forward_sampling.py [hunk 5]
```diff
@@ -113,24 +109,24 @@ def sample_prior_predictive(
...
     ...     model2(), sample_shape=(20,), sample_from_observed=False
     ... )
-    >>> np.allclose(np.mean(prior_samples["model2/y"]), 1)
+    >>> np.allclose(np.mean(prior_samples.prior_predictive["model2/y"]), 1)
     True
 
     Furthermore, this has consequences at the shape level of the drawn samples
     >>> prior_samples["model2/y"].shape
-    (20, 10)
+    (20, 20)
```
[the remark this answers is not in the document]
Comment (aloctavodia) [reply pymc4/forward_sampling.py]: you are right, this is wrong. I will fix it.

File: pymc4/forward_sampling.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (lucianopaz) [reply pymc4/forward_sampling.py]: Does arviz impose the notion of chains in prior predictive sampling? I think this can be very dangerous in models with many random variables, some that are scalar and some that are multidimensional. The scalar ones might end up with an extra "chain" axis, but what happens to the multidimensional ones? Do they get an extra axis as well or are they left as they were, trusting that some sample shape axis indicated the chain membership? In any case, it will lead to potential and unexpected broadcasting problems down the line.

File: tests/test_forward_sampling.py (hunk 3 above)
Comment (lucianopaz) [reply tests/test_forward_sampling.py]: Do we always need the chain axis in an inference data object?

Comment (ColCarroll) [reply tests/test_forward_sampling.py]: yeah, for now we do - we can avoid this if we do the translation inside pymc4

File: pymc4/forward_sampling.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ColCarroll) [reply pymc4/forward_sampling.py]: arviz assumes you always indicate the number of chains, and that it is always a single scalar.

**Comment by ColCarroll** [comment]
Oy, while playing around with examples, the other problem I saw was that this makes posterior predictive sampling a bit awkward, right? I would love to be able to do 

```python
trace = pm.sample(model())
ppc_trace = pm.sample_posterior_predictive(trace)
```

or even 

```python
trace = pm.sample(model())
pm.sample_posterior_predictive(trace, in_place=True)
```
but right now `sample_posterior_predictive` expects a dictionary of numpy arrays (with a certain shape!)

File: pymc4/forward_sampling.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (lucianopaz) [reply pymc4/forward_sampling.py]: I understand that the chain dimension comes out naturally from `sample` and thus it makes its way into the posterior predictive samples. But I think that it isn't right to impose it to the prior predictive samples.
Furthermore, one cannot make assumptions on whether it is correct to always add a dimension to the samples from the prior. For example, in smc with multiple chains, one has to explicitly draw independent samples from the prior for each chain. It wouldn't be right to later add an extra dimension to the resulting samples.

File: pymc4/inference/utils.py [hunk 6]
```diff
@@ -31,23 +31,49 @@ def initialize_sampling_state(
...
+    posterior_predictive : dict
+    inplace : bool
 
     Returns
     -------
-    arviz.data.inference_data.InferenceData
+    ArviZ's InferenceData object
     """
+    if trace is not None and isinstance(trace, dict):
+        trace = {k: np.swapaxes(v.numpy(), 1, 0) for k, v in trace.items() if "/" in k}
```
[the remark this answers is not in the document]
Comment (aloctavodia) [reply pymc4/inference/utils.py]: The `if "/" in k` prevents adding to the posterior group the observed data. observed data  it's added to its own group. I have not found an example where this becomes problematic. Could you provide a short example of such scenario?

File: pymc4/inference/utils.py [hunk 7]
```diff
@@ -31,23 +31,49 @@ def initialize_sampling_state(
...
+    if sample_stats is not None and isinstance(sample_stats, dict):
+        sample_stats = {k: v.numpy().T for k, v in sample_stats.items()}
+    if prior_predictive is not None and isinstance(prior_predictive, dict):
+        prior_predictive = {k: v[np.newaxis] for k, v in prior_predictive.items() if "/" in k}
```
[the remark this answers is not in the document]
Comment (aloctavodia) [reply pymc4/inference/utils.py]: I think is OK to remove it from here. I have the vague memory of having to add this for some reason, but probably I am just confused.

**Comment by aloctavodia** [comment]
I think by merging this we will have a better opportunity to discuss what objects the sampling functions of PyMC4 should return, and also to discuss and potentially improve the InferenceData Schema from ArviZ, that is still evolving.

File: pymc4/inference/utils.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (lucianopaz) [reply pymc4/inference/utils.py]: The `if "/" in k` doesn't prevent adding the observed data to the posterior group. The observed data just isn't returned by `sample` because it only collects the values of transformed, untransformed and deterministics variables. In the following model:

```python
@pm.model:
def model()
    mu = yield pm.Normal(loc=0, scale=1, name="mu")
    observed = yield pm.Normal(loc=mu, scale=1, name="observed", observed=np.zeros(10))
    return observed
```

Evaluating the model will give you the following variables: "model/mu", "model/observed", "model". The last "model" variable is stored as a deterministic, and that means that it will be returned in the `sample`ed trace. The `if "/" in k` means that the deterministic variable named "model" will be dropped. It doesn't look like a big deal here, but it isn't transparent to the user. 

Furthermore, if the users pass a coroutine directly to the prior predictive sampler, no "model/" will be prepended, and all the root level variables will be dropped.

[the remark this answers is not in the document]
Comment (aloctavodia) [reply pymc4/inference/utils.py]: "model" is dropped here, because observed data has it's own group and it does not belong to the posterior group. see https://github.com/pymc-devs/pymc4/pull/199/files#diff-2eb557123ba49614ee42a55d2dae7fdfR160

I have removed the `if "/" in k` from the prior_predictive.

**Comment by ColCarroll** [comment]
Do the notebooks still run here? I see a huge diff on the baseball notebook!

I'm ok with this if tests and notebooks are passing, but won't merge until @lucianopaz gives the "ok".

**Comment by aloctavodia** [comment]
The notebooks run know as part of the test #205, so they should be OK, right?

**Comment by ColCarroll** [comment]
Yep! Just rechecked the baseball notebook (which was -1800 lines in the diff!) and it seems as though your images are just smaller than whoever executed it last (me?).

**Comment by lucianopaz** [comment]
I just feel a bit conflicted with the fact that if we merge this, we will changing the samples dimensions to add a chains dimension that 
 matches arviz's schema.
I'll approve and merge this PR as it is nice to have a tight integration, but I would really like it if arviz's could entertain the possibility of having an empty chains dimension. This would be useful for prior predictive samples and possibly also for samples generated from a fitted VI model.

**Comment by aloctavodia** [comment]
Great! We will discuss all your concern with the ArviZ team.
